dockerproxy/main.go

```diff
@@ -470,6 +472,32 @@ func authorizeRequest(appName, authToken string) bool {
 func extendDeadline() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		log.Infof("extendDeadline called with user agent: %s", r.UserAgent())
+
+		di, err := disk.GetInfo("/data")
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			log.Errorf("failed to check /data: %s", err)
+			return
+		}
+
+		if  di.Free <= uint64(1 * gb) {
+			w.WriteHeader(http.StatusInsufficientStorage)
+
+			free := float64(di.Free) / float64(gb)
+			s := fmt.Sprintf("remaining disk space (%.2fGB) is too low", free)
+			w.Write([]byte(s))
+			return
+		}
+
+		percentUsed := (float64(di.Total-di.Free) / float64(di.Total))
+		if percentUsed >= pruneThresholdUsedPercent {
+			w.WriteHeader(http.StatusInsufficientStorage)
+
+			s := fmt.Sprintf("remaining disk space (%.2f%%) is too low", percentUsed)
+			w.Write([]byte(s))
+			return
+		}
```

Comment:
Is there anything we can do instead of extending the deadline so builds keep working? Can we prune something from Docker, even during a running build?

Reply:
We already prune on startup if the disk space is low.

It is better to _succeed_ than fail here, even if it's slower.

Reply:
Okay. The second commit is adding prune() in extendDeadline.